### PR TITLE
Connect Keccak circuit and table

### PR DESCRIPTION
**Pull request by ChihChengLiang**
Presentation for #712

### CODE REVIEW

**Comment by Brechtpd** [comment]
Looks good. One small thing is that `is_final` (or now `is_enabled`) is set to true on the 1st row but it doesn't necessarily contain valid data on that row (it's set to true to there to make the circuit easier). So the correct way would be to do `not(q_first) * is_final` for `is_enabled`. However I was thinking of forcing the first row to contain the keccak hash of no data (`0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470`) which is very easy/cheap to do and this way the first row isn't a special case anymore (at least from an external point of view). What do you think?

**Review by CPerezz — APPROVED** [review]
LGTM! Looks nice @ChihChengLiang thanks for this work!!!!

**Comment by roynalnaruto** [comment]
@ChihChengLiang not sure why the SHA3 opcode tests are failing.

Are we missing a `dev_load` from the EVM circuit?

**Comment by CPerezz** [comment]
It seems that some tests are not passing due to the latest changes. We should review that carefully.

File: zkevm-circuits/src/evm_circuit.rs [hunk 1]
```diff
@@ -310,11 +310,7 @@ pub mod test {
             config
                 .copy_table
                 .load(&mut layouter, &self.block, self.block.randomness)?;
-            config.keccak_table.load(
-                &mut layouter,
-                &self.block.sha3_inputs,
-                self.block.randomness,
-            )?;
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit.rs:317]: Shouldn't this be changed from `load` to `dev_load` (and not completely removed)?

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit.rs:317]: Agree. Let me edit it! :smile:

**Review by roynalnaruto — APPROVED** [review]
Looks good to me!

**Comment by CPerezz** [comment]
Looks like there's still a failure in the keccak_bit approach. Currently investigatig..

File: zkevm-circuits/src/keccak_circuit/keccak_bit.rs [hunk 2]
```diff
@@ -644,20 +647,16 @@ impl<F: Field> KeccakBitConfig<F> {
...
-        }
+        self.keccak_table.assign_row(
+            region,
+            offset,
+            [
+                F::from(row.is_final),
+                F::from(row.length as u64),
+                row.data_rlc,
+                row.hash_rlc,
+            ],
```
[the remark this answers is not in the document]
Comment (roynalnaruto) [reply zkevm-circuits/src/keccak_circuit/keccak_bit.rs:658]: Yep, from the EVM circuit, we can see that the lookup order is:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/connect-keccak-table/zkevm-circuits/src/evm_circuit/table.rs#L359-L368

So this change should fix it!

[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply zkevm-circuits/src/keccak_circuit/keccak_bit.rs]: Thanks, @Brechtpd @roynalnaruto, and @CPerezz! Looks like it is the case and the tests passed!

**Pull request merged**